orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/providers/gce/GceDeployStagePreProcessor.java

```diff
@@ -0,0 +1,152 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.orca.clouddriver.pipeline.providers.gce;
+
+import static com.netflix.spinnaker.orca.kato.pipeline.support.ResizeStrategySupport.getSource;
+
+import com.google.common.collect.ImmutableList;
+import com.netflix.spinnaker.orca.clouddriver.pipeline.providers.aws.ApplySourceServerGroupCapacityStage;
+import com.netflix.spinnaker.orca.clouddriver.pipeline.providers.aws.CaptureSourceServerGroupCapacityTask;
+import com.netflix.spinnaker.orca.clouddriver.pipeline.servergroup.ResizeServerGroupStage;
+import com.netflix.spinnaker.orca.clouddriver.pipeline.servergroup.strategies.AbstractDeployStrategyStage;
+import com.netflix.spinnaker.orca.clouddriver.pipeline.servergroup.strategies.DeployStagePreProcessor;
+import com.netflix.spinnaker.orca.clouddriver.pipeline.servergroup.support.TargetServerGroupResolver;
+import com.netflix.spinnaker.orca.kato.pipeline.strategy.Strategy;
+import com.netflix.spinnaker.orca.kato.pipeline.support.ResizeStrategy;
+import com.netflix.spinnaker.orca.kato.pipeline.support.StageData;
+import com.netflix.spinnaker.orca.pipeline.model.Stage;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+@Component
+public class GceDeployStagePreProcessor implements DeployStagePreProcessor {
+  private final ApplySourceServerGroupCapacityStage applySourceServerGroupSnapshotStage;
+  private final ResizeServerGroupStage resizeServerGroupStage;
+  private final TargetServerGroupResolver targetServerGroupResolver;
+
+  @Autowired
+  public GceDeployStagePreProcessor(
+      ApplySourceServerGroupCapacityStage applySourceServerGroupCapacityStage,
+      ResizeServerGroupStage resizeServerGroupStage,
+      TargetServerGroupResolver targetServerGroupResolver) {
+    this.applySourceServerGroupSnapshotStage = applySourceServerGroupCapacityStage;
+    this.resizeServerGroupStage = resizeServerGroupStage;
+    this.targetServerGroupResolver = targetServerGroupResolver;
+  }
+
+  @Override
+  public List<StepDefinition> additionalSteps(Stage stage) {
+    return ImmutableList.of(
+        new StepDefinition(
+            "snapshotSourceServerGroup", CaptureSourceServerGroupCapacityTask.class));
+  }
+
+  @Override
+  public List<StageDefinition> beforeStageDefinitions(Stage stage) {
+    StageData stageData = stage.mapTo(StageData.class);
+    if (shouldPinSourceServerGroup(stageData.getStrategy())) {
```

Comment:
Feel free to disagree and ignore, but I'd be inclined to flip this if statement. IMO it's usually easier to read if the bulk of the logic is at the bottom of the method instead of the middle.

```
if (!shouldPinSourceServerGroup(stageData.getStrategy())) {
  return ImmutableList.of();
}

// etc
```

Hot IntelliJ tip: If you agree, you can do this by hitting alt-enter on `if` and selecting `invert if condition`.

Reply:
I agreed and I hit alt-enter!